### ISSUE TITLE
Can't simulate multimedia keys

### ISSUE THREAD

**Issue by Jont828** [autokey/autokey#127]
## Classification: 
Performance/Feature
## Reproducibility: 
Always
## Summary: 
Cannot make shortcuts to multimedia keys (i.e. play/pause, which is the key XF86AudioPlay)
## Steps to Reproduce: 
XF86AudioPlay, XF86AudioLowerVolume, XF86AudioRaiseVolume, XF86AudioPrev, and XF86AudioNext is not in the list of special keys. I have tried system.exec_command("xdotool key XF86AudioPlay") and system.exec_command("xdotool key --clearmodifiers XF86AudioPlay"), but that did not work, either.
## Expected Results: 
Multimedia key presses are simulated
## Actual Results: 
Nothing happens
## Version: 0.94.0
Installed via:  AUR
Distro:  Arch
## Notes
Tried system.exec_command("xdotool key XF86AudioPlay") and system.exec_command("xdotool key --clearmodifiers XF86AudioPlay"), but that did not work. Occasionally, the key press will be registered, but it's very unreliable. However, **it seems to work after checking "Always prompt before executing this script and clicking the prompt, so I guess I have found a temporary workaround**. Nevertheless, it would be best to be able to use keyboard.send_keys() to do it as it has worked for my non-multimedia shortcuts.

**Comment by Conceptron** [comment]
Try using `time.sleep(.1)` before `system.exec_command()`, also call `system.exec_command("command", getOutput=False)` otherwise autokey will wait for the output of the command. Doing this registers the keypress most of the time

**Comment by Jont828** [comment]
Thanks for the reply! So is this how I should call it, then? I put that in, but it's pretty laggy.
```
time.sleep(.1)
system.exec_command("xdotool key XF86AudioRaiseVolume", getOutput=Faddlse)
```

**Comment by Conceptron** [comment]
Yes, you could try decreasing the sleep duration just above the point where it doesn't register the key anymore to decrease the lagginess. As you have suggested, it is best to add special keys for these media keys, as `xdotool` seems to be conflicting with `Autokey` in some way.

**Comment by luziferius** [comment]
It does work natively with AutoKey. But you may run into an issue that your receiving application discards programmatically generated key presses. (For details see #132.)
I tried sending `XF86AudioRaiseVolume` and it worked in the event tester application `xev`, but the KDE volume meter ignores the key press…
Then I tried a script sending `XF86Back` to firefox, and it worked. So this depends on the receiving application.

To send a multimedia key, you need the (decimal number) key code and use the special `<code###>` syntax in keyboard.send_keys(), replacing `###` with your key code.
To obtain it, use `xev`and lookout for the `keycode` attribute in the third row:
```
KeyPress event, serial 51, synthetic NO, window 0x7c00001,
    root 0x157, subw 0x0, time 12345365, (100,-20), root:(1829,9),
    state 0x0, keycode 172 (keysym 0x1008ff14, XF86AudioPlay), same_screen YES,
    XLookupString gives 0 bytes: 
    XmbLookupString gives 0 bytes: 
    XFilterEvent returns: False

```
Alternatively, use the AutoKey `Set Hotkey` dialog (the one that is used to assign hotkeys to scripts and phrases). Start recording a key combination, then press the desired key you want to send.
AutoKey shows the key using it’s special key syntax. You can note it down and cancel the dialog.
Then use the noted string to send your key.

Example script:
```
keyboard.send_keys("<code172>")  # sends an `XF86AudioPlay` key press.
```

**Comment by Conceptron** [comment]
Which application executes the effects of keys like `XF86AudioPlay` in Ubuntu 16.04? `keyboard.send_keys("<code172>")` didn't work for me, `xdotool key 172` worked though. 

Could be because (as you pointed out in #132), `xdotool` has `synthetic no` and `send_keys` has `synthetic yes` in `xev`.

Is there any way of making that application accept synthetic keys?

**Comment by luziferius** [comment]
About which program causes the key press event to be handled: I’m not sure. The application (e.g. the volume meter) might listen on key presses on its own like autokey does, or use some system service. The latter method is used in KDE as far as I know.

About the synthetic keys: It depends. `xterm` for example can be configured to accept those, others might or might not have a configuration switch.
Well, if `xdotool` outputs 'non-synthetic' key presses, maybe we can adopt its method. (Reading its source code and adopt it, if possible) Or write an `xdotool` backend, that uses this tool to send keys instead of the xlib method currently in use.

**Comment by jeebak** [comment]
This thread was very helpful for me to solve my issue. Particularly:

> About which program causes the key press event to be handled: I’m not sure. The application (e.g. the volume meter) might listen on key presses on its own like autokey does, or use some system service. The latter method is used in KDE as far as I know.

That lead me to this:
 https://unix.stackexchange.com/questions/342554/how-to-enable-my-keyboards-volume-keys-in-xfce/342555
... and since my system uses `pulseaudio`, what ultimately worked for me was:
```
system.exec_command("pactl -- set-sink-volume 0 -5%", getOutput=False)
```

**Comment by luziferius** [comment]
@jeebak This looks like a really clean solution. (But it only works for controlling pulseaudio directly.)
Instead of going the route `keyboard` → `AutoKey` → `emulated keyboard` → `some other background service` → `pulseaudio`, your attempt is the most direct possible: `keyboard` → `AutoKey` → `pulseaudio`.

The whole system is event driven, so other volume meters will still see the volume change and adjust their sliders, if visible.

**Comment by tlan16** [comment]
Somehow `time.sleep(.1)` did not work for me (Arch, KDE), but this worked:

```
time.sleep(0.5)
system.exec_command("xdotool key XF86AudioNext")
```

**Comment by AdrianSkar** [comment]
> Somehow `time.sleep(.1)` did not work for me (Arch, KDE), but this worked:
> 
> ```
> time.sleep(0.5)
> system.exec_command("xdotool key XF86AudioNext")
> ```

Same here (Arch, KDE), working properly with:
```
time.sleep(.5)
system.exec_command("xdotool key XF86AudioPlay", getOutput=False)
```

**Comment by strugk** [comment]
`time.sleep(0.15)
system.exec_command("xdotool key XF86AudioNext")`
Works at my with delay reduced to 0.15 - but fails to work at 0.1 - semi-works with unpredictable results in the range 0.11-0.14

**Comment by josephj11** [comment]
@strugk Thanks for the feedback.

Every system runs a bit differently, so timing and delays usually need tuning.

**Comment by Elliria** [comment]
It probably won't work, but might be worth a try to do two `time.sleep(.05)` delays back-to-back to total up to the desired .1-second delay.

**Comment by josephj11** [comment]
@Elliria Why? It would probably be imperceptibly longer than one .1 second delay, just adding the time it takes Python to interpret and run another instruction.

**Comment by Elliria** [comment]
It was just a wild guess since `time.sleep(0.1)` doesn't work. I suppose one could get around that imperceptibly-longer time by doing `time.sleep(0.04)` and `time.sleep(0.05)` or maybe `time.sleep(0.04)` and `time.sleep(0.04)` to get it to add up to a .1-second delay. Either way, I figured it probably wouldn't work, but might be worth a try. Whatever is stopping the .1-second delay from working might be capable of being tricked and might never see those shorter delays coming.
